src/main/java/sg/ncl/MainController.java

```diff
@@ -232,40 +232,51 @@ public String loginSubmit(
-            jwtTokenString = responseEntity.getBody().toString();
+        HttpEntity<String> request = new HttpEntity<>(headers);
+        restTemplate.setErrorHandler(new MyResponseErrorHandler());
+
+        try {
+            response = restTemplate.exchange(properties.getSioAuthUrl(), HttpMethod.POST, request, String.class);
+            jwtTokenString = response.getBody().toString();
         } catch (Exception e) {
-            // TODO: handle different types of error
-            // case1: invalid login
-            loginForm.setErrorMsg("Login failed: Invalid email/password.");
+            logger.warn("Error connecting to authentication service to validate login details");
+            loginForm.setErrorMsg(ERR_SERVER_OVERLOAD);
             return "login";
         }
 
-        if (jwtTokenString != null || !jwtTokenString.isEmpty()) {
-            JSONObject tokenObject = new JSONObject(jwtTokenString);
-            String token = tokenObject.getString("token");
-            id = tokenObject.getString("id");
-
-            logger.info("login success with token id {}", id);
-            AUTHORIZATION_HEADER = "Bearer " + token;
-
-            session.setAttribute("sessionLoggedEmail", loginForm.getLoginEmail());
-            session.setAttribute("id", id);
-
-            return "redirect:/dashboard";
-        } else {
-            // case1: invalid login
-            loginForm.setErrorMsg("Login failed: Invalid email/password.");
-            return "login";
+        try {
+            if (RestUtil.isError(response.getStatusCode())) {
+                MyErrorResource error = objectMapper.readValue(jwtTokenString, MyErrorResource.class);
+                loginForm.setErrorMsg("Login failed: Invalid email/password.");
+                logger.warn("login failed with username {}", loginForm.getLoginEmail());
+                return "login";
+            } else if (jwtTokenString != null || !jwtTokenString.isEmpty()) {
+                JSONObject tokenObject = new JSONObject(jwtTokenString);
+                String token = tokenObject.getString("token");
+                id = tokenObject.getString("id");
+
+                AUTHORIZATION_HEADER = "Bearer " + token;
+
+                session.setAttribute("sessionLoggedEmail", loginForm.getLoginEmail());
+                session.setAttribute("id", id);
+
+                logger.info("login success with username: {}, token id: {}", loginForm.getLoginEmail(), id);
+                return "redirect:/dashboard";
+            } else {
+                // case1: invalid login
+                loginForm.setErrorMsg("Login failed: Invalid email/password.");
+                logger.warn("login failed with unknown response code with username {}", loginForm.getLoginEmail());
+                return "login";
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS00112)